### PR TITLE
Support ARM64 atomic instructions, workaround for PAC instructions

### PR DESCRIPTION
**Pull request by peternguyen93**
Hi team,

I have modified some code path to enable and workaround some features in order to emulate arm64e binary on iOS/macOS.

### CODE REVIEW

**Comment by aquynh** [comment]
Nice. Did you backport from newer version of Qemu?

**Comment by peternguyen93** [comment]
I forked from unicorn dev branch, these features were available on Qemu but unicorn didn't enable aarch64 atomic instructions and PAC instructions.

**Comment by aquynh** [comment]
Which Qemu version are you referring to?

**Comment by peternguyen93** [comment]
hi @aquynh I just modified code in qemu directory in unicorn project.

File: qemu/target/arm/unicorn_aarch64.c [hunk 1]
```diff
@@ -438,6 +438,18 @@ static int arm64_cpus_init(struct uc_struct *uc, const char *cpu_model)
         return -1;
     }
 
+    // enable AARCH64 atomic instruction
```
Comment (wtdcode) [inline qemu/target/arm/unicorn_aarch64.c:441]: This looks like a dirty fix.

File: samples/sample_arm64.c [hunk 2]
```diff
@@ -89,6 +89,89 @@ static void test_arm64_mem_fetch(void)
...
+    if (err) {
+        printf("Failed on uc_emu_start() with error returned: %u\n", err);
+    }
+
+    uc_reg_read(uc, UC_ARM64_REG_X16, &x9);
+    uc_mem_read(uc, x9, (void *)&output, 8);
+    printf("Output: %llx\n", (uint64_t)output);
+}
+
+static void test_arm64_pacda(void)
```
Comment (wtdcode) [inline samples/sample_arm64.c:135]: Could you move these functions to `tests/unit`? Also I'm not sure the reason you need this workaround.

Comment (peternguyen93) [reply samples/sample_arm64.c:135]: I added this to test my patch is actually work or not, do you want me to remove this ?

Comment (peternguyen93) [reply samples/sample_arm64.c:135]: hi @wtdcode Sure, I added 2 function tests here to debug and make sure I correctly enable these instructions.

**Comment by aquynh** [comment]
In the sample you can have a new variable x16, instead of reusing x9, which
is confused.

In addition, you should fork the dev branch, not the master branch, instead
to make pull req

**Comment by peternguyen93** [comment]
Do I need to remove this pull request and re-create it with dev branch?

**Comment by aquynh** [comment]
> Do I need to remove this pull request and re-create it with dev branch?

yes, please.

**Pull request closed without merging**